File: 1-project-setup/dotnet-client/consumer.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Confluent.Kafka;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using System.Text.Json;


namespace DotnetClient
{
    public class Consumer
    {
        private readonly IConfiguration _configuration;
        private readonly string _bootstrapServer;
        private readonly bool _isLocal;
        private readonly string? _topic;
        private readonly ILogger<Consumer> _logger;
        private readonly CancellationTokenSource _cts;

        public Consumer(IConfiguration configuration, ILogger<Consumer> logger)
        {
            _configuration = configuration;
            _logger = logger;
            _topic = _configuration["Kafka:Topic"];
            _bootstrapServer = _configuration["Kafka:BootstrapServers"] ?? "localhost:9092";

            // Check for local environments
            _isLocal = _bootstrapServer.Contains("localhost") ||
                       _bootstrapServer.Contains("127.0.0.1") ||
                       _bootstrapServer.Contains("kafka:");

            _cts = new CancellationTokenSource();

            // Setup Ctrl+C handler
            Console.CancelKeyPress += (_, e) =>
            {
                e.Cancel = true;
                _cts.Cancel();
            };
        }

        public async Task Consume(int messageLimit = 10)
        {
            if (string.IsNullOrEmpty(_topic))
            {
                _logger.LogWarning("⚠️ No topic specified");
                return;
            }

            // Configure Kafka consumer
            var consumerConfig = new ConsumerConfig
            {
                BootstrapServers = _bootstrapServer,
                GroupId = Environment.GetEnvironmentVariable("GROUP_ID") ?? "dotnet-consumer-group-id",
                AutoOffsetReset = AutoOffsetReset.Earliest,
                ClientId = Environment.GetEnvironmentVariable("CLIENT_ID") ?? "dotnet-client"
            };

            // Configure security based on environment
            if (_isLocal)
            {
                consumerConfig.SecurityProtocol = SecurityProtocol.Plaintext;
                _logger.LogInformation("🐳 Using PLAINTEXT protocol for local broker");
                _logger.LogInformation("🔄 Using topic {Topic} - will be auto created if it doesn't exist", _topic);
            }
            else
            {
                consumerConfig.SecurityProtocol = SecurityProtocol.SaslSsl;
                consumerConfig.SaslMechanism = SaslMechanism.Plain;
                consumerConfig.SaslUsername = Environment.GetEnvironmentVariable("CC_API_KEY");
                consumerConfig.SaslPassword = Environment.GetEnvironmentVariable("CC_API_SECRET");
                _logger.LogInformation("☁️ Using SASL_SSL protocol for cloud broker");
            }

            // Verify Kafka setup before proceeding
            try
            {
                bool exists = await VerifyKafkaSetup(consumerConfig, _topic, _isLocal);
                if (!exists && !_isLocal)
                {
                    _logger.LogError("Topic '{Topic}' does not exist. Please create it first", _topic);
                    return;
                }
                _logger.LogInformation("✅ Connected to Kafka");
            }
            catch (Exception e)
            {
                _logger.LogError(e, "Kafka connection error: {Message}", e.Message);
                return;
            }

            _logger.LogInformation("Using plain JSON deserialization (no Schema Registry)");

            _logger.LogInformation("Listening on topic: '{Topic}'...", _topic);

            try
            {
                using (var consumer = new ConsumerBuilder<string, string>(consumerConfig)
                    .Build())
                {
                    consumer.Subscribe(_topic);
                    // Increased delay to allow for topic auto-creation in local environments
                    if (_isLocal)
                    {
                        await Task.Delay(5000);
                    }

                    int count = 0;

                    while (count < messageLimit && !_cts.Token.IsCancellationRequested)
                    {
                        try
                        {
                            var consumeResult = consumer.Consume(TimeSpan.FromSeconds(1));

                            if (consumeResult == null)
                                continue;

                            count++;
                            try
                            {
                                // Deserialize JSON message
                                var transaction = JsonSerializer.Deserialize<TransactionRecord>(consumeResult.Message.Value);
                                
                                // Format the output as requested
                                string jsonOutput = consumeResult.Message.Value;
                                _logger.LogInformation("Message {Current}/{Total}: {JsonData}", count, messageLimit, jsonOutput);
                            }
                            catch (JsonException ex)
                            {
                                _logger.LogError(ex, "Error deserializing message: {Message}", ex.Message);
                                _logger.LogDebug("Raw message: {RawMessage}", consumeResult.Message.Value);
                            }
                        }
                        catch (ConsumeException e)
                        {
                            _logger.LogError(e, "Error: {Reason}", e.Error.Reason);
                            if (e.InnerException != null)
                            {
                                _logger.LogError("Inner Exception: {Message}", e.InnerException.Message);
                            }
                        }
                        catch (Exception e)
                        {
                            _logger.LogError(e, "❌ Error processing message: {Message}", e.Message);
                            if (e.InnerException != null)
                            {
                                _logger.LogError("❌ Inner Exception: {Message}", e.InnerException.Message);
                            }
                        }
                    }

                    consumer.Close();
                    _logger.LogInformation("✅ Consumed {Count} messages from {Topic}", count, _topic);
                }
            }
            catch (OperationCanceledException)
            {
                _logger.LogInformation("Interrupted by user");
            }
        }

        private Task<bool> VerifyKafkaSetup(ConsumerConfig config, string? topic, bool isLocal)
        {
            if (isLocal || string.IsNullOrEmpty(topic))
            {
                return Task.FromResult(true);
            }

            try
            {
                var adminClientConfig = new AdminClientConfig
                {
                    BootstrapServers = config.BootstrapServers,
                    SecurityProtocol = config.SecurityProtocol,
                    SaslMechanism = config.SaslMechanism,
                    SaslUsername = config.SaslUsername,
                    SaslPassword = config.SaslPassword
                };

                using (var adminClient = new AdminClientBuilder(adminClientConfig).Build())
                {
                    var metadata = adminClient.GetMetadata(TimeSpan.FromSeconds(5));
                    return Task.FromResult(metadata.Topics.Exists(t => t.Topic == topic));
                }
            }
            catch
            {
                return Task.FromResult(false);
            }
        }

    }
}

File: 2-project-modified/go-client/consumer/consumer.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"strings"
	"syscall"

	"github.com/confluentinc/confluent-kafka-go/v2/kafka"
	"github.com/joho/godotenv"
)

type Transaction struct {
	TransactionId   string  `json:"TransactionId" avro:"TransactionId"`
	AccountNumber   string  `json:"AccountNumber" avro:"AccountNumber"`
	Amount          float64 `json:"Amount" avro:"Amount"`
	Currency        string  `json:"Currency" avro:"Currency"`
	Timestamp       string  `json:"Timestamp" avro:"Timestamp"`
	TransactionType string  `json:"TransactionType" avro:"TransactionType"`
	Status          string  `json:"Status" avro:"Status"`
}


func isLocalKafka(bootstrapServer string) bool {
	return strings.Contains(bootstrapServer, "localhost") ||
		strings.Contains(bootstrapServer, "127.0.0.1") ||
		strings.Contains(bootstrapServer, "kafka:")
}

// verifyKafkaSetup checks Kafka connection and topic
func verifyKafkaSetup(conf kafka.ConfigMap, topic string, isLocal bool) error {
	if isLocal {
		fmt.Println("✅ Kafka connection assumed available for local environment")
		return nil
	}

	adminClient, err := kafka.NewAdminClient(&conf)
	if err != nil {
		return fmt.Errorf("Kafka configuration error: %w", err)
	}
	defer adminClient.Close()

	metadata, err := adminClient.GetMetadata(&topic, false, 5000)
	if err != nil {
		return fmt.Errorf("Kafka connection error: %w", err)
	}

	if _, exists := metadata.Topics[topic]; !exists {
		return fmt.Errorf("Topic '%s' does not exist", topic)
	}

	return nil
}

// handleJSONMessage processes a message as plain JSON
func handleJSONMessage(value []byte, messageCount, maxMessages int) {
	var txn Transaction
	if err := json.Unmarshal(value, &txn); err != nil {
		fmt.Printf("❌ Error processing message: %v\n", err)
		fmt.Printf("Binary message, size: %d bytes\n", len(value))
		return
	}

	// Print message details
	fmt.Printf("Consumed message %d/%d: ", messageCount, maxMessages)

	// Print the raw JSON
	jsonOutput, _ := json.Marshal(txn)
	fmt.Printf("%s\n", string(jsonOutput))
}

func main() {
	err := godotenv.Load()
	if err != nil {
		log.Printf("Warning: Error loading .env file: %v", err)
	}

	bootstrapServer := os.Getenv("CC_BOOTSTRAP_SERVER")
	topic := os.Getenv("CC_TOPIC")
	groupID := os.Getenv("GROUP_ID")
	if groupID == "" {
		groupID = "go-client-consumer-group"
	}

	conf := kafka.ConfigMap{
		"bootstrap.servers": bootstrapServer,
		"group.id":          groupID,
		"auto.offset.reset": "earliest",
	}

	// Detect if using local Kafka
	isLocal := isLocalKafka(bootstrapServer)

	if isLocal {
		conf["security.protocol"] = "PLAINTEXT"
		conf["client.id"] = os.Getenv("CLIENT_ID")
		fmt.Println("🐳 Using PLAINTEXT protocol for local broker")
	} else {
		conf["security.protocol"] = "SASL_SSL"
		conf["sasl.mechanisms"] = "PLAIN"
		conf["sasl.username"] = os.Getenv("CC_API_KEY")
		conf["sasl.password"] = os.Getenv("CC_API_SECRET")
		conf["client.id"] = os.Getenv("CLIENT_ID")
		fmt.Println("☁️ Using SASL_SSL protocol for cloud broker")
	}

	// Verify Kafka setup before proceeding
	if err := verifyKafkaSetup(conf, topic, isLocal); err != nil {
		fmt.Printf("❌ Kafka configuration error - Exiting: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("✅ Connected to Kafka (%s)\n", bootstrapServer)

	fmt.Println("🔄 Using plain JSON deserialization (no Schema Registry)")

	// Create consumer
	consumer, err := kafka.NewConsumer(&conf)
	if err != nil {
		fmt.Printf("Failed to create consumer: %s\n", err)
		os.Exit(1)
	}
	defer consumer.Close()

	// Subscribe to topic
	err = consumer.SubscribeTopics([]string{topic}, nil)
	if err != nil {
		fmt.Printf("Failed to subscribe to topic %s: %s\n", topic, err)
		os.Exit(1)
	}

	fmt.Printf("Listening on topic: %s\n", topic)

	// Set up a channel for handling signals to terminate
	sigchan := make(chan os.Signal, 1)
	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)

	// Process messages or terminate
	run := true
	messageCount := 0
	maxMessages := 10

	for run && messageCount < maxMessages {
		select {
		case sig := <-sigchan:
			fmt.Printf("Caught signal %v: terminating\n", sig)
			run = false
		default:
			ev := consumer.Poll(100)
			if ev == nil {
				continue
			}

			switch e := ev.(type) {
			case *kafka.Message:
				messageCount++
				value := e.Value

				// Process as JSON
				handleJSONMessage(value, messageCount, maxMessages)

			case kafka.Error:
				fmt.Fprintf(os.Stderr, "ERROR: %v\n", e)
				if e.Code() == kafka.ErrAllBrokersDown {
					run = false
				}
			default:
				// Ignore other event types
			}
		}
	}

	fmt.Printf("✅ Consumed %d messages from %s\n", messageCount, topic)
	fmt.Println("Consumer closed")
}
